src/protocols/secure_channel/CASESession.cpp

```diff
@@ -129,6 +132,164 @@ static constexpr ExchangeContext::Timeout kExpectedLowProcessingTime    = System
+                                        Platform::Delete<WorkHelper>);
+        if (ptr)
+        {
+            ptr->mWeakPtr = ptr; // used by `ScheduleWork`
+        }
+        return ptr;
+    }
+
+    // Schedule the work after configuring the data.
+    // If lifetime is managed, the helper shares management while work is outstanding.
+    CHIP_ERROR ScheduleWork()
+    {
+        if (!mSession || !mWorkCallback || !mAfterWorkCallback)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+        mStrongPtr  = mWeakPtr.lock(); // set in `Create`
+        auto status = DeviceLayer::PlatformMgr().ScheduleBackgroundWork(WorkHandler, reinterpret_cast<intptr_t>(this));
+        if (status != CHIP_NO_ERROR)
+        {
+            mStrongPtr.reset();
+        }
+        return status;
+    }
+
+    // Cancel the work, by clearing the associated session.
+    void CancelWork() { mSession.store(nullptr); }
+
+private:
+    // Handler for the work callback.
+    static void WorkHandler(intptr_t arg)
+    {
+        WorkHelper * helper = reinterpret_cast<WorkHelper *>(arg);
+        bool cancel         = false;
+        VerifyOrExit(helper->mSession.load(), ;); // cancelled by `CancelWork`?
+        helper->mStatus = helper->mWorkCallback(helper->mData, cancel);
+        VerifyOrExit(!cancel, ;);                 // canceled by `mWorkCallback`?
+        VerifyOrExit(helper->mSession.load(), ;); // cancelled by `CancelWork`?
+        SuccessOrExit(DeviceLayer::PlatformMgr().ScheduleWork(AfterWorkHandler, reinterpret_cast<intptr_t>(helper)));
+        return;
+    exit:
+        helper->mStrongPtr.reset();
+    }
+
+    // Handler for the after work callback.
+    static void AfterWorkHandler(intptr_t arg)
+    {
+        // Since this runs in the main CHIP thread, the session shouldn't be otherwise used (messages, timers, etc.)
+        WorkHelper * helper = reinterpret_cast<WorkHelper *>(arg);
+        if (auto * session = helper->mSession.load())
+        {
+            (session->*(helper->mAfterWorkCallback))(helper->mData, helper->mStatus);
+        }
+        helper->mStrongPtr.reset();
+    }
+
+private:
+    // Lifetime management: `ScheduleWork` sets `mStrongPtr` from `mWeakPtr`.
+    std::weak_ptr<WorkHelper> mWeakPtr;
+
+    // Lifetime management: `ScheduleWork` sets `mStrongPtr` from `mWeakPtr`.
+    std::shared_ptr<WorkHelper> mStrongPtr;
+
+    // Associated session, cleared by `CancelWork`.
+    std::atomic<CASESession *> mSession;
+
+    // Work callback, called by `WorkHandler`.
+    WorkCallback mWorkCallback;
+
+    // After work callback, called by `AfterWorkHandler`.
+    AfterWorkCallback mAfterWorkCallback;
+
+    // Return value of `mWorkCallback`, passed to `mAfterWorkCallback`.
+    CHIP_ERROR mStatus;
+
+public:
+    // Data passed to `mWorkCallback` and `mAfterWorkCallback`.
+    DATA mData;
+};
+
+struct CASESession::SendSigma3Data
+{
+    FabricTable * fabricTable;
```

Comment:
It doesn't seem like it is going to be safe to access the fabricTable in a background thread without locking.

Reply:
It's tricky to mutex the fabric table during background operations. And we can't copy signing keys, and there's not a good way to proxy to them (it can almost be done, but they can be externally owned outside the fabric table, so there's a risk they could be deleted out from under use), and with operational keystore being fully abstract/virtual, there's not a lot we can do there.

So, I just made it background if using the keystore (and it can deal with mutex etc.) and not background if using the legacy method (which is typically simpler anyways). Maybe if we eventually remove the legacy method, then we can make it always background.

Reply:
>  (and it can deal with mutex etc.)

That's a change in the keystore API, right?  At a bare minimum that needs to be documented in the keystore header...

Reply:
Yes. I just added a virtual function which gives OperationalKeystore ability to declare if it's OK with this. If not, CASE won't background the opkeypair signing.